# Data Analysis with Python

This Python Notebook is to perform Data Analysis using Python

First thing first, import all required library.



# Import Library

In [ ]:
import pandas as pd

In [ ]:
pd.__version__

'1.1.5'

# Load Data

To load the dataset, we have various options based on the source format. Here we have the source format as csv, thus we will be using Pandas `read_csv()` method, which will read the csv, and load the entire dataset into DataFrame.

So incase if you are not sure what DataFrame is -- 

    Pandas DataFrame is a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns). Arithmetic operations align on both row and column labels. It can be thought of as a dict-like container for Series objects. This is the primary data structure of the Pandas.

In [ ]:
df = pd.read_csv(r'/content/drive/MyDrive/Colab Notebooks/DataSets/Sample_Weather_Data.csv')

The general thumb rule in any Machine Learning Problem is to know your data, understand your data. Get the basic information out of it such as 
* how many data points or observations our dataset has?
* does it has any missing values?
* how many columns / features our dataset has?
* what is the data type of each column?
* and so on.

Based on this next step is decided.

You may call this with any name you like such as EDA (Exploratory Data Analysis); or Understanding Data; or Exploring Data; Performing Data Analysis etc

The Purpose here is to get basic understanding of the dataset, and the problem we are trying to solve.

Without wasting the time, lets see what methods Python and Pandas has provided to us.

 # 1. How to analyze Data Frame?

# head()
The very first this after loading the data is to see how our data looks like, for this we have to have a glance of first few data points, for this we have a method called `head()`. 

Pandas `head()` method is used to return top n (5 by default) rows of a data frame or series.

In [ ]:
df.head()

Date/Time  Temp_C  ...  Press_kPa               Weather
0  1/1/2012 0:00    -1.8  ...     101.24                   Fog
1  1/1/2012 1:00    -1.8  ...     101.24                   Fog
2  1/1/2012 2:00    -1.8  ...     101.26  Freezing Drizzle,Fog
3  1/1/2012 3:00    -1.5  ...     101.27  Freezing Drizzle,Fog
4  1/1/2012 4:00    -1.5  ...     101.23                   Fog

[5 rows x 8 columns]

We could see the first or top 5 records, as we did not provided any value, it is defaulted to 5, thus returning first 5 records. In Pandas DataFrame are indexed from 0, so we have 0 to 4 which is total of 5.

In [ ]:
df.head(8)

Date/Time  Temp_C  ...  Press_kPa               Weather
0  1/1/2012 0:00    -1.8  ...     101.24                   Fog
1  1/1/2012 1:00    -1.8  ...     101.24                   Fog
2  1/1/2012 2:00    -1.8  ...     101.26  Freezing Drizzle,Fog
3  1/1/2012 3:00    -1.5  ...     101.27  Freezing Drizzle,Fog
4  1/1/2012 4:00    -1.5  ...     101.23                   Fog
5  1/1/2012 5:00    -1.4  ...     101.27                   Fog
6  1/1/2012 6:00    -1.5  ...     101.29                   Fog
7  1/1/2012 7:00    -1.4  ...     101.26                   Fog

[8 rows x 8 columns]

We got first 8 records.

# tail()

Similar to `head()` method, pandas has `tail()` method to show the last n rows.

Pandas `tail()` method is used to return last n (5 by default) rows of a data frame or series.

In [ ]:
df.tail()

Date/Time  Temp_C  ...  Press_kPa  Weather
8779  12/31/2012 19:00     0.1  ...     100.13     Snow
8780  12/31/2012 20:00     0.2  ...     100.03     Snow
8781  12/31/2012 21:00    -0.5  ...      99.95     Snow
8782  12/31/2012 22:00    -0.2  ...      99.91     Snow
8783  12/31/2012 23:00     0.0  ...      99.89     Snow

[5 rows x 8 columns]

We could see the last 5 records, as we did not provided any value, it is defaulted to 5, thus returning last 5 records.

In [ ]:
df.tail(2)

Date/Time  Temp_C  ...  Press_kPa  Weather
8782  12/31/2012 22:00    -0.2  ...      99.91     Snow
8783  12/31/2012 23:00     0.0  ...      99.89     Snow

[2 rows x 8 columns]

Here we could see last 2 records.

Also the last record is at index 8783

# shape
It shows the total number of rows (data points or observations) and total number of columns of the dataframe.

Pandas has provided `.size` and `.ndim` along with `.shape` which are used to return size dimensions,and shape respectively.  of data frames and series.

In [ ]:
df.shape

(8784, 8)

So there are total of 8784 records or data points with 8 columns.

From here itself we could say that the dataframe is of 2 dimensional and the size of dataframe is 8784 * 8 = 70272

Same information can be checked using pandas `.size` and `.ndim` methods.

In [ ]:
df.size

70272

In [ ]:
df.shape[0] * df.shape[1]

70272

So we could see that the size if just the multiplication of rows and columns, which alternatively written as `df.shape[0] * df.shape[1]`.

In [ ]:
df.ndim

2

See we already knew that our dataframe is of 2 dimensional, usually ML problems will have dataframe of 2 dimensional array, and when we are working on Images (specially gray or colored images) we do have 3 dimensional arrays. Please do explore if interested.

What if we need to verify the dimension of an individual column or a series, no problem, lets do it.

In [ ]:
df['Wind Speed_km/h'].ndim

1

So when we are working with individual columns from the dataframe it will be converted into series, and the series are 1 dimensional, as we seen above.

Pandas `Series` is a one-dimensional labelled array capable of holding any data type (integers, strings, floating point numbers, Python objects, etc.). It has to be remembered that unlike Python lists, a Series will always contain data of the same type.


# index
`index` attribute provides the index or range index of the dataframe.

In our sample dataset we know that the index has started from zero (0) which we validated using `.head()` and the last row has index as 8783, which we validated using `.tail()` method.

Same is verified using `.index`.

In [ ]:
df.index

RangeIndex(start=0, stop=8784, step=1)

Oops, here we got the RangeIndex which says that it started from 0 agreed, but howcome it stops at 8784 where as we knew that the last indexed row is 8783.

Here is the catch, with the index range, last value is excluded. Thus when it says it stop as 8784, the actual index is ended at 8784 - 1 = 8783.

# columns
Next is `.columns` attribute, which is to get all the column names in the data frame.

In [ ]:
df.columns

Index(['Date/Time', 'Temp_C', 'Dew Point Temp_C', 'Rel Hum_%',
       'Wind Speed_km/h', 'Visibility_km', 'Press_kPa', 'Weather'],
      dtype='object')

Here will get the list of Columns from the data frame.

# dtypes

Pandas `.dtypes` attribute is used to get the data type of each column in the DataFrame.



In [ ]:
df.dtypes 

Date/Time            object
Temp_C              float64
Dew Point Temp_C    float64
Rel Hum_%             int64
Wind Speed_km/h       int64
Visibility_km       float64
Press_kPa           float64
Weather              object
dtype: object

Above result shows the data type of each and every columns in the data frame.

What is we are just interested in any one or specific set of columns. Lets see that as well.

In [ ]:
df['Wind Speed_km/h'].dtypes

dtype('int64')

Here is the data type of individual columns, for example I have used `Wind Speed_km/h`, similarly one could use any column.

In [ ]:
df[['Wind Speed_km/h', 'Weather']].dtypes

Wind Speed_km/h     int64
Weather            object
dtype: object

When checking the data type of multiple columns make sure to provide them in double [[ ]] as shown above.

# unique()
Pandas `unique()` method is used to show all unique values for a given column. Remember that this `unique()` method can be applied on a single column only, and not on the whole data frame.

In [ ]:
df['Wind Speed_km/h'].unique()

array([ 4,  7,  6,  9, 15, 13, 20, 22, 19, 24, 30, 35, 39, 32, 33, 26, 44,
       43, 48, 37, 28, 17, 11,  0, 83, 70, 57, 46, 41, 52, 50, 63, 54,  2])

In [ ]:
df.Weather.unique()

array(['Fog', 'Freezing Drizzle,Fog', 'Mostly Cloudy', 'Cloudy', 'Rain',
       'Rain Showers', 'Mainly Clear', 'Snow Showers', 'Snow', 'Clear',
       'Freezing Rain,Fog', 'Freezing Rain', 'Freezing Drizzle',
       'Rain,Snow', 'Moderate Snow', 'Freezing Drizzle,Snow',
       'Freezing Rain,Snow Grains', 'Snow,Blowing Snow', 'Freezing Fog',
       'Haze', 'Rain,Fog', 'Drizzle,Fog', 'Drizzle',
       'Freezing Drizzle,Haze', 'Freezing Rain,Haze', 'Snow,Haze',
       'Snow,Fog', 'Snow,Ice Pellets', 'Rain,Haze', 'Thunderstorms,Rain',
       'Thunderstorms,Rain Showers', 'Thunderstorms,Heavy Rain Showers',
       'Thunderstorms,Rain Showers,Fog', 'Thunderstorms',
       'Thunderstorms,Rain,Fog',
       'Thunderstorms,Moderate Rain Showers,Fog', 'Rain Showers,Fog',
       'Rain Showers,Snow Showers', 'Snow Pellets', 'Rain,Snow,Fog',
       'Moderate Rain,Fog', 'Freezing Rain,Ice Pellets,Fog',
       'Drizzle,Ice Pellets,Fog', 'Drizzle,Snow', 'Rain,Ice Pellets',
       'Drizzle,Snow,Fog', 

When working with columns we could use it in two ways.
* df.Weather --> When the column name does not have any spaces in between.
* df['Weather'] --> always recommended to use. Easy to readibility.

Above unique() method returned all unique values for the given column.

# nunique()
Pandas `.nunique()` method is to show the total number of unique values in each column. As mentioned this can be applied on the whole data frame as well as on a single or subset of columns.

In [ ]:
df.nunique()

Date/Time           8784
Temp_C               533
Dew Point Temp_C     489
Rel Hum_%             83
Wind Speed_km/h       34
Visibility_km         24
Press_kPa            518
Weather               50
dtype: int64

Here I have used the nunique with the entire data frame. Now lets see how to use it with single and subset of columns.

In [ ]:
df.Weather.nunique()

50

In [ ]:
df[['Wind Speed_km/h', 'Weather']].nunique()

Wind Speed_km/h    34
Weather            50
dtype: int64

Above two code snippet shows that total numbe of unique values in weather column is 50, and when check as a subset of columns it shown same, alsong with for all provided columns.

# count()
Pandas `count()` method is used to show the total numbe of non-null in each column.

`count()` method can be applied on a single column as well as on whole data frame.

In [ ]:
df.count()

Date/Time           8784
Temp_C              8784
Dew Point Temp_C    8784
Rel Hum_%           8784
Wind Speed_km/h     8784
Visibility_km       8784
Press_kPa           8784
Weather             8784
dtype: int64

From this we can easily say that there are no null or Nan values we our data frame contain 8784 rows, and each column also has same 8784 values (irrespective of duplicate or unique).

In [ ]:
df.Weather.count()

8784

In [ ]:
df[['Wind Speed_km/h', 'Weather']].count()

Wind Speed_km/h    8784
Weather            8784
dtype: int64

As mentioned count() method can also be used on single column or on subset of columns which is shown above.

# isnull() & notnull()
While making a Data Frame from a csv file, many blank columns are imported as null value into the Data Frame which later creates problems while operating that data frame. 

Sometime our data-set itself contains missing data points, which when imported into Data Frame will be converted into null values or NaN.

Pandas `isnull()` and `notnull()` methods are used to check and manage NULL values in a data frame.


In [ ]:
df.isnull()

Date/Time  Temp_C  Dew Point Temp_C  ...  Visibility_km  Press_kPa  Weather
0         False   False             False  ...          False      False    False
1         False   False             False  ...          False      False    False
2         False   False             False  ...          False      False    False
3         False   False             False  ...          False      False    False
4         False   False             False  ...          False      False    False
...         ...     ...               ...  ...            ...        ...      ...
8779      False   False             False  ...          False      False    False
8780      False   False             False  ...          False      False    False
8781      False   False             False  ...          False      False    False
8782      False   False             False  ...          False      False    False
8783      False   False             False  ...          False      False    False

[8784 rows x 8 columns]

`isnull()` will return True os there is any NaN value present in the value, else will be false.

However seeing such a big dataset with True or False will be very tedious, to make it simpler we could aggregate it and check if there is any null value, using sum() on top of isnull() as below.

In [ ]:
df.isnull().sum()

Date/Time           0
Temp_C              0
Dew Point Temp_C    0
Rel Hum_%           0
Wind Speed_km/h     0
Visibility_km       0
Press_kPa           0
Weather             0
dtype: int64

From `counts()` method we already knew that our dataset has no null value which we also confirmed with `isnull().sum()` method.

Opposite to `isnull()` we do have `notnull()` method which will return True for every NON-NULL value and False for a null value.

Here we well we could use `.sum()` aggregate function to get the count.

In [ ]:
df.notnull()

Date/Time  Temp_C  Dew Point Temp_C  ...  Visibility_km  Press_kPa  Weather
0          True    True              True  ...           True       True     True
1          True    True              True  ...           True       True     True
2          True    True              True  ...           True       True     True
3          True    True              True  ...           True       True     True
4          True    True              True  ...           True       True     True
...         ...     ...               ...  ...            ...        ...      ...
8779       True    True              True  ...           True       True     True
8780       True    True              True  ...           True       True     True
8781       True    True              True  ...           True       True     True
8782       True    True              True  ...           True       True     True
8783       True    True              True  ...           True       True     True

[8784 rows x 8 columns]

In [ ]:
df.notnull().sum()

Date/Time           8784
Temp_C              8784
Dew Point Temp_C    8784
Rel Hum_%           8784
Wind Speed_km/h     8784
Visibility_km       8784
Press_kPa           8784
Weather             8784
dtype: int64

# value_counts()

Pandas `value_count()` is another method to show all the unique values with their count for a given column. This method is applied on a single columns only.

The resulting object will be in descending order so that the first element is the most frequently-occurring element. Excludes NA values by default.

In [ ]:
df['Weather'].value_counts()

Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Haze                                         16
Thunderstorms,Rain Showers                   16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle                        

Above I mentioned that `value_counts()` is applied on single column, but what if we have a requirement to get the uniuqe value count for subset of columns or on whole data frame.

There are ways with which we could solve the problem.
* Loop over each column.
* use of `apply` and pass series.value_counts()
* etc.

In [ ]:
df[['Wind Speed_km/h', 'Weather']].apply(pd.Series.value_counts)

Wind Speed_km/h  Weather
0                                                  309.0      NaN
2                                                    2.0      NaN
4                                                  474.0      NaN
6                                                  609.0      NaN
7                                                  677.0      NaN
...                                                  ...      ...
Thunderstorms,Moderate Rain Showers,Fog              NaN      1.0
Thunderstorms,Rain                                   NaN      3.0
Thunderstorms,Rain Showers                           NaN     16.0
Thunderstorms,Rain Showers,Fog                       NaN      3.0
Thunderstorms,Rain,Fog                               NaN      1.0

[84 rows x 2 columns]

# info()
Pandas `info()` method is used to show a concise summary of a DataFrame. This method displays basic information about a DataFrame including the 
* index range   
* column(s) dtypes, 
* Count of non-null values 
* memory usage

All these details we have seens above using individual pandas method or attributes, however with `.info()` method we could get in one single line of code.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8784 entries, 0 to 8783
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Date/Time         8784 non-null   object 
 1   Temp_C            8784 non-null   float64
 2   Dew Point Temp_C  8784 non-null   float64
 3   Rel Hum_%         8784 non-null   int64  
 4   Wind Speed_km/h   8784 non-null   int64  
 5   Visibility_km     8784 non-null   float64
 6   Press_kPa         8784 non-null   float64
 7   Weather           8784 non-null   object 
dtypes: float64(4), int64(2), object(2)
memory usage: 549.1+ KB


# describe()
Last but not the least is pandas `describe()` method, which is used to view some basic statisticial details like percentile, mean, standard deviation, etc of a dataframe or a series of numeric values.

Remember the `describe()` method by-default return the detail for numeric value, however we could check the details for `object` type as well, in this the method will return a different output which includes total count of values, unique value; top and frequency of occurence from each object type column.

In [ ]:
df.describe()

Temp_C  Dew Point Temp_C  ...  Visibility_km    Press_kPa
count  8784.000000       8784.000000  ...    8784.000000  8784.000000
mean      8.798144          2.555294  ...      27.664447   101.051623
std      11.687883         10.883072  ...      12.622688     0.844005
min     -23.300000        -28.500000  ...       0.200000    97.520000
25%       0.100000         -5.900000  ...      24.100000   100.560000
50%       9.300000          3.300000  ...      25.000000   101.070000
75%      18.800000         11.800000  ...      25.000000   101.590000
max      33.000000         24.400000  ...      48.300000   103.650000

[8 rows x 6 columns]

If we observe the output carefully then we could easily identify that the columns `Weather` and `Date/Time` is not display, as the data type form them is `object`.

Lets display them now.

In [ ]:
df.describe(include=['object'])

Date/Time       Weather
count               8784          8784
unique              8784            50
top     10/18/2012 10:00  Mainly Clear
freq                   1          2106

If need to view for all columns including all data types, use the below code.

In [ ]:
df.describe(include=['object', 'int', 'float'])

Date/Time       Temp_C  ...    Press_kPa       Weather
count               8784  8784.000000  ...  8784.000000          8784
unique              8784          NaN  ...          NaN            50
top     10/18/2012 10:00          NaN  ...          NaN  Mainly Clear
freq                   1          NaN  ...          NaN          2106
mean                 NaN     8.798144  ...   101.051623           NaN
std                  NaN    11.687883  ...     0.844005           NaN
min                  NaN   -23.300000  ...    97.520000           NaN
25%                  NaN     0.100000  ...   100.560000           NaN
50%                  NaN     9.300000  ...   101.070000           NaN
75%                  NaN    18.800000  ...   101.590000           NaN
max                  NaN    33.000000  ...   103.650000           NaN

[11 rows x 8 columns]

`describe()` method has two more parameters apart from `include` which are `percentile` and `exclude`. I recommend to please check them, and explore it.

Now as we are familiar with some of the basic methods and attributes, lets try to answer the following questions.

Please make a note that there are multiple ways to solve any problem, and here we are just focusing on one or two ways. Please do write in comments if you have used any other method, so that others including me will be benefited with that code.

# Question 1 : Determine how many times the Weather was "Clear"?



## approach 1 : value_counts()
Use `value_counts()` method to display all the counts, and just pick the count for "Clear". 


In [ ]:
df['Weather'].value_counts()

Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Haze                                         16
Thunderstorms,Rain Showers                   16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle                        

From above output we could say that the "Clear" weather was 1326 times.

However this is not a good way to show the result.

## Approach 2 : Filter
Filter the required data, to get the count of specifc value from a column.

In [ ]:
df['Weather'] == 'Clear'

0       False
1       False
2       False
3       False
4       False
        ...  
8779    False
8780    False
8781    False
8782    False
8783    False
Name: Weather, Length: 8784, dtype: bool

In [ ]:
df[df['Weather'] == 'Clear']['Weather'].count()

1326

Perfect, this is what we are looking for. So 1326 times the weather was "Clear".

If want the sub-set of data with "Clear" weather, then just use the code `df[df['Weather'] == 'Clear']`

In [ ]:
df[df['Weather'] == 'Clear']

Date/Time  Temp_C  ...  Press_kPa  Weather
67      1/3/2012 19:00   -16.9  ...     101.74    Clear
114     1/5/2012 18:00    -7.1  ...     100.71    Clear
115     1/5/2012 19:00    -9.2  ...     100.80    Clear
116     1/5/2012 20:00    -9.8  ...     100.83    Clear
117     1/5/2012 21:00    -9.0  ...     100.83    Clear
...                ...     ...  ...        ...      ...
8646   12/26/2012 6:00   -13.4  ...     102.47    Clear
8698  12/28/2012 10:00    -6.1  ...     101.27    Clear
8713   12/29/2012 1:00   -11.9  ...     101.31    Clear
8714   12/29/2012 2:00   -11.8  ...     101.33    Clear
8756  12/30/2012 20:00   -13.8  ...     101.52    Clear

[1326 rows x 8 columns]

# Question 2 : On which days the Wind Speed was exactly 4 km per hour, and what was the Weather on that day.

In [ ]:
df[df['Wind Speed_km/h'] == 4][['Date/Time', 'Weather', 'Wind Speed_km/h']]

Date/Time       Weather  Wind Speed_km/h
0        1/1/2012 0:00           Fog                4
1        1/1/2012 1:00           Fog                4
96       1/5/2012 0:00          Snow                4
101      1/5/2012 5:00          Snow                4
146      1/7/2012 2:00        Cloudy                4
...                ...           ...              ...
8768   12/31/2012 8:00  Snow Showers                4
8769   12/31/2012 9:00          Snow                4
8770  12/31/2012 10:00      Snow,Fog                4
8772  12/31/2012 12:00          Snow                4
8773  12/31/2012 13:00          Snow                4

[474 rows x 3 columns]

Here you go, there are 474 records when the wind speed was 4 km per hr, and we have the date time with the weather for that duration.

# Question 3 : Get the list of columns which are of object type.

## Approach 1: using `.dtype` and `.columns`

In [ ]:
[c for c in df.columns if df[c].dtype in ['object']]

['Date/Time', 'Weather']

## Approach 2 : using alternate method as `.select_dtypes()` and attribute `.columns`.

In [ ]:
list(df.select_dtypes(['object']).columns)

['Date/Time', 'Weather']